File: Examples/CSharp/x-IMU3-Examples/Examples/NetworkDiscovery.cs
using System;

namespace Ximu3Examples
{
    class NetworkDiscovery
    {
        public NetworkDiscovery()
        {
            if (Helpers.YesOrNo("Use async implementation?") == true)
            {
                Ximu3.NetworkDiscovery discovery = new Ximu3.NetworkDiscovery(NetworkDiscoveryEvent);
                Helpers.Wait(-1);
                discovery.Dispose();
            }
            else
            {
                Ximu3.DiscoveredNetworkDevice[] devices = Ximu3.NetworkDiscovery.Scan(2000);
                Console.WriteLine("Devices updated (" + devices.Length + " devices available)");
                PrintDevices(devices);
            }
        }

        private void NetworkDiscoveryEvent(Object sender, Ximu3.NetworkDiscoveryEventArgs args)
        {
            Console.WriteLine("Discovered " + args.devices.Length + " devices");
            PrintDevices(args.devices);
        }

        private void PrintDevices(Ximu3.DiscoveredNetworkDevice[] devices)
        {
            foreach (Ximu3.DiscoveredNetworkDevice device in devices)
            {
                Console.WriteLine(device.DeviceName + " - " +
                    device.SerialNumber + ", RSSI: " +
                    device.Rssi + "%, Battery: " +
                    device.Battery + "%, " +
                    device.Status + ", " +
                    device.TcpConnectionInfo + ", " +
                    device.UdpConnectionInfo);
                // Console.WriteLine(device); // alternative to above
            }
        }
    }
}

File: Examples/Python/network_discovery.py
import helpers
import ximu3


def print_devices(devices):
    for device in devices:
        print(device.device_name + " - " +
              device.serial_number + ", RSSI: " +
              str(device.rssi) + "%, Battery: " +
              str(device.battery) + "%, " +
              device.status + ", " +
              device.tcp_connection_info.to_string() + ", " +
              device.udp_connection_info.to_string())
        # print(device.to_string())  # alternative to above


def callback(devices):
    print("Devices updated (" + str(len(devices)) + " devices available)")
    print_devices(devices)


if helpers.yes_or_no("Use async implementation?"):
    _ = ximu3.NetworkDiscovery(callback)
    helpers.wait(10)
else:
    devices = ximu3.NetworkDiscovery.scan(2000)
    print("Discovered " + str(len(devices)) + " devices")
    print_devices(devices)
